presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExpressionEquivalence.java

```diff
@@ -292,6 +292,33 @@ else if (rightValue == null) {
                 return Integer.compare(((InputReferenceExpression) left).getField(), ((InputReferenceExpression) right).getField());
             }
 
+            if (left instanceof LambdaDefinitionExpression) {
+                LambdaDefinitionExpression leftLambda = (LambdaDefinitionExpression) left;
+                LambdaDefinitionExpression rightLambda = (LambdaDefinitionExpression) right;
+
+                return ComparisonChain.start()
+                        .compare(
+                                leftLambda.getArgumentTypes(),
+                                rightLambda.getArgumentTypes(),
+                                new ListComparator<>(Comparator.comparing(Object::toString)))
```

Comment:
What happens when type or variable names differ in case?

Reply:
comparison results would be either -1 or 1. Given that: `LambdaDefinitionExpression#equals` is:
```
        ...
        LambdaDefinitionExpression that = (LambdaDefinitionExpression) o;
        return Objects.equals(argumentTypes, that.argumentTypes) &&
                Objects.equals(arguments, that.arguments) &&
                Objects.equals(body, that.body);
```
this is a correct behaviour

Reply:
It would be appropriate to consider `a -> a + 1` and `b -> b + 1` as equal. Given how planning and optimization works, this kind of change could (at least in concept, most likely also in practice) happen. However, I think it would be reasonable to leave that as a TODO for now. My understanding is that this PR is release blocking.